src/senaite/core/adapters/localroles.py

```diff
@@ -0,0 +1,89 @@
+# -*- coding: utf-8 -*-
+#
+# This file is part of SENAITE.CORE.
+#
+# SENAITE.CORE is free software: you can redistribute it and/or modify it under
+# the terms of the GNU General Public License as published by the Free Software
+# Foundation, version 2.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+# FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
+# details.
+#
+# You should have received a copy of the GNU General Public License along with
+# this program; if not, write to the Free Software Foundation, Inc., 51
+# Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+#
+# Copyright 2018-2022 by it's authors.
+# Some rights reserved, see README and LICENSE.
+
+import six
+from bika.lims import api
+from bika.lims import logger
+from borg.localrole.default_adapter import DefaultLocalRoleAdapter
+from senaite.core.interfaces import IDynamicLocalRoles
+from zope.component import getAdapters
+
+
+class DynamicLocalRoleAdapter(DefaultLocalRoleAdapter):
+    """Gives additional Member local roles based on current user and context
+    This enables giving additional permissions on items out of the user's
+    current traverse path
+    """
+
+    _roles_in_context = {}
+
+    def getRolesInContext(self, context, principal_id):
+        """Returns the dynamically calculated 'local' roles for the given
+        principal and context
+        @param context: context to calculate roles for the given principal
+        @param principal_id: User login id
+        @return List of dynamically calculated local-roles for user and context
+        """
+        if not api.is_object(context):
+            # We only apply dynamic local roles to valid objects
+            return []
+
+        # This function is called a lot within same request, do some cache
+        context_uid = api.get_uid(context)
+        roles = self._roles_in_context.get(context_uid, {})
+        if principal_id in roles:
+            return roles.get(principal_id)
+
+        # Look for adapters
+        roles = []
+        path = api.get_path(context)
+        adapters = getAdapters((context,), IDynamicLocalRoles)
+        for name, adapter in adapters:
+            local_roles = adapter.getRoles(principal_id)
+            logger.info(u"{}::{}::{}: {}".format(name, path, principal_id,
+                                                 repr(local_roles)))
+            roles.extend(local_roles)
```

Comment:
Instead of extending, you might want to use directly a `set`. This would omit the conversion done in the next step.